MRG: Refactor Xdawn

kingjr opened the pull request:
* Refactor: Xdawn and XdawnTransformer classes, now both in `mne.preprocessing.xdawn`
* FIX: bug weird channels picking in tests and Xdawn class
* FIX: unstable unit test
* FIX: remove undocumented attributes (e.g. `Xdawn.exclude`), add and document necessary ones (`Xdawn.overlap_`, `evokeds_`, `.event_id_`)
* ENH: `_xdawn_fit` private function for more explicit args.
* FIX: `XdawnTransformer` devoid of any fancy objects (Evoked / dictionary etc), and only uses numpy

cc @kaichogami @alexandrebarachant @dengemann

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,332 @@
+
+    # Get prototype events
+    if events is not None:
+        evokeds, toeplitzs = _least_square_evoked(epochs, events, tmin, sfreq)
```
Btw this (that is, `_least_square_evoked`) is where an encoder model backbone would fit in.
`toeplitz` is an encoder's X, `evokeds` are the coefs of the fitted model/the estimated rERPs.

(As @alexandrebarachant has noted a few times, originally, XDawn is applied to continuous data, and this internally re-creates the continuous data.)

kingjr replied:
Good, I was actually unclear what toeplitz was :/

Can you suggest a snippet to incorporate the encoding model here in a subsequent PR?

jona-sassenhagen replied:
With the current code, it would be something like

```
from mne.stats.regression import _prepare_rerp_preds, _make_evokeds
from mne.preprocessing.xdawn import _construct_signal_from_epochs
from sklearn.linear_model import Ridge
raw_, events_ = _construct_signal_from_epochs(events, epochs.info["sfreq"], tmin, tmax, epochs.get_data())  # function would have to be modified to return events)
toeplitzs, conds, cond_length, tmin_s, tmax_s = _prepare_rerp_preds(
    raw_.n_times, raw_.info["sfreq"], events_, event_id, tmin, tmax)
coefs = Ridge(alpha=0).fit(toeplitzs, raw_._data).coefs_
evokeds = _make_evokeds(coefs, conds, cond_length, tmin_s, tmax_s, raw_.info)
```

Hm. Doesn't look that hard actually.
I think I'll try finally doing this (1 year after I opened the issue) once this one is merged.

To build the evokeds, you need some information about the delays/estimated response functions per condition to extract the data from coefs. So if you use a different mechanism to build X/toeplitzs, that will have to be taken care of.

jona-sassenhagen replied:
_prepare_rerp_preds would have to be modified to optionally return the unstacked X, because XDawn uses the stacked X in _least_square_evoked, but XDawn also needs them unstacked later. Alternatively, I think X can be memory efficiently reconstructed on the fly, not sure.

jona-sassenhagen replied:
Okay, I think this would work:

from mne.stats.regression import _prepare_rerp_preds, _make_evokeds
from mne.preprocessing.xdawn import _construct_signal_from_epochs
from sklearn.linear_model import Ridge
from scipy import sparse

# function would have to be modified to return events)
raw_, events_ = _construct_signal_from_epochs(
    events, epochs.info["sfreq"], tmin, tmax, epochs.get_data())

X, conds, cond_length, tmin_s, tmax_s = _prepare_rerp_preds(
    raw_.n_times, raw_.info["sfreq"], events_, event_id, tmin, tmax, stack=False)
coefs = Ridge(alpha=0).fit(sparse.hstack(X), raw_._data).coefs_
evokeds = _make_evokeds(coefs, conds, cond_length, tmin_s, tmax_s, raw_.info)
toeplitzs = [t_.toarray() for t_ in X]

Would be a 2 line change to _make_evokeds, and allow us to remove about 100 lines of code from XDawn and around 20 lines of code from linear_regression_raw tests.

kingjr replied:
> Would be a 2 line change to _make_evokeds, and allow us to remove about 100 lines of code from XDawn and around 20 lines of code from linear_regression_raw tests.

Nice, shall I try it in this PR, or does it require some other code?

jona-sassenhagen replied:
Let me do it. I've been postponing this for much too long.

kingjr replied:
Can you open an Issue so that we don't forget, I won't handle this in this PR

jona-sassenhagen replied:
There already is one: https://github.com/mne-tools/mne-python/issues/2332

kingjr commented:
@Eric89GXL I'm not sure I did the deprecation correctly could you check?

@alexandrebarachant I incorporated the overlap correction here. Do you think it should be added direclty to pyRiemann?

@kaichogami @dengemann @jona-sassenhagen  Could you give it a review API wise?

kingjr commented:
here is  a snippet to test the new API:

```
import matplotlib.pyplot as plt
import numpy as np
from mne.preprocessing import XdawnTransformer
xd = XdawnTransformer(n_components=3)
n_trial, n_chan, n_time = 100, 20, 10
X = np.random.randn(n_trial, n_chan, n_time)
y = np.random.randint(0, 2, n_trial)
xd.fit(X, y)
Xt = xd.transform(X)
X_denoised = xd.inverse_transform(Xt)
plt.matshow(X_denoised.mean(0))
plt.show()
```

agramfort commented on examples/preprocessing/plot_xdawn_denoising.py:
```diff
@@ -68,13 +68,23 @@
+# Plot in sensor space
+X_denoised = xd.inverse_transform(Xt)
+epochs_denoised = epochs.copy()
+epochs_denoised._data[test] = X_denoised
```
I prefer if you use EpochsArray rather than hacking private attributes.

kingjr replied:
yes. I'm actually wondering whether I shouldn't just plot the data, since we're not using any particular topography, might as well keep it simple, WDYT?

agramfort commented on examples/preprocessing/plot_xdawn_denoising.py:
```diff
@@ -34,7 +34,7 @@
                  Epochs)
 from mne.datasets import sample
-from mne.preprocessing import Xdawn
+from mne.preprocessing import XdawnTransformer
```
wait I though we said we have Xdawn in preprocessing for denoising and that handles overlap. And have XdawnTransformer in decoding that is lighter and does not support overlap. So no need to pass events to XdawnTransformer in fit method.

kingjr replied:
Good point. We're indeed facing 2 distinct issues that I mixed in this PR:

1. The XdawnTransformer did not take overlapping events. 

I changed the code to have an `events` param at fit time: `xd.fit(X, y, events=events)`, but I'm wondering whether this is such a great idea. It's sklearn compatible but not conventional (yet?) . So, there is a risk that this API is not going to be sustainable.

2. The Xdawn class is an API mess. 

I modularized its core computation into separate private functions. 

The XdawnTransformer now does the *very* same computations but keeps a strict and simple API closer to sklearn than to `mne.ICA` or `mne.GAT`. Specifically, it doesn't have an `apply` method, takes numpy arrays, does not keep `mne.Evoked` objects in its attributes etc.

Eric89GXL commented:
I think that the decorator can work directly on a class like that. But just
try instantiating the class, that will tell you if it's working.

kaichogami commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,332 @@
-    return evokeds
+    n_components : int (default 2)
+        The number of components to decompose the signals.
+    reg : float | str | None (default None)
```
Brackets around `None`.

kingjr replied:
what do you mean @kaichogami , the rest of the doc doesn't seem to have bracket?

kaichogami replied:
It does, but @dengemann suggested it should be written in normal text and I changed it accordingly there. Or am I missing something here?

kaichogami commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,332 @@
+            evokeds.append(np.mean(epochs[y == c, :, :], axis=0))
+            toeplitzs.append(1.)
+
+    filters = []
```
Explicit initialization of list.

kaichogami commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,332 @@
+        The epochs class.
+    n_components : int (default 2)
+        The number of components to decompose the signals signals.
+    reg : float | str | None (default None)
```
Brackets around `None`.

kaichogami commented:
I think its a good idea to keep one `Xdawn` algorithm instead of dividing them into two. However I am not sure about the third parameter in `fit`. Didn't we write the `Xdawntransformer` instead of changing `Xdawn` as we were not passing `events` which resulted `overlap` not being supported ?

agramfort commented:
>
> I think its a good idea to keep one Xdawn algorithm instead of dividing
> them into two. However I am not sure about the third parameter in fit.
> Didn't we write the Xdawntransformer instead of changing Xdawn as we were
> not passing events which resulted overlap not being supported ?
>
well you can have the same code but expose it using private class or
functions as 2 classes dedicated to distinct usage. You don't have to
duplicate any code.

jona-sassenhagen commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -77,6 +78,7 @@ def test_xdawn_fit():
     assert_raises(ValueError, xd.fit, epochs)
     # fit with baseline correction and ovverlapp correction should throw an
     # error
+    # XXX This is a buggy test, the epochs here don't overlap
```
FWIW the EEGLAB example data has overlapping events.

kingjr replied:
Thx. We have to simulate however, else the tests are slower, no?

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,332 @@
-    evoked_data = dict()
-    for idx, eid in enumerate(event_id):
-        evoked_data[eid] = all_evokeds[idx].T
+    return all_evokeds, toeplitz_mat
```
Docstring calls it "evoked_data"

jona-sassenhagen commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -77,6 +78,7 @@ def test_xdawn_fit():
     assert_raises(ValueError, xd.fit, epochs)
     # fit with baseline correction and ovverlapp correction should throw an
```
ovverlapp -> overlap

kingjr commented:
Overall, we could indeed keep a single `Xdawn` class. The motivation for creating `XdawnTransformer` was the necessity for a deprecation cycle so as to be able to change the API. 

Specifically, 
*  we needed `xd.filters_` and `xd.patterns_` to be stored as numpy arrays instead of dicts (as the dicts depends on unnecessary `event_id` params), and
* we do not want to store `xd.evokeds_` which involve complex `Evoked` objects, and are uncessary for the Xdawn computations.

The latter points needs not creating a new class, but I believe that the former does.

Overall, I think we have multiple non-exclusive options

1. Keep a `Xdawn` and `XdawnTransformer` class, which will respectively be deprecated, and become the standard API.

2. Make `XdawnTransformer` accepts both epochs and numpy arrays: `xd.fit(epochs) # overlap` and `xd.fit(X, y) #no overlap`. The former would allow extracting the events, but not the latter.

3. Make XdawnTransformer accepts numpy arrays only, and pass an additional `events` param at fitting time (the current solution), although this is not really conventional in sklearn.

cc @agramfort @kaichogami @dengemann

kingjr commented on examples/preprocessing/plot_xdawn_denoising.py:
```diff
@@ -70,11 +69,11 @@
 xd.fit(epochs)
 
-# Denoise epochs
+# Denoise epochs. Note that it outputs a numpy array
```
woops

kingjr commented on examples/preprocessing/plot_xdawn_denoising.py:
```diff
@@ -70,11 +69,11 @@
 xd = Xdawn(n_components=2, signal_cov=signal_cov)
 
-# Fit xdawn
+# Fit xdawn and pass events to correct for overlapping epochs
```
woops

kingjr commented:
I tried to successively implement three different API, and came to the conclusion to stick to the 3rd one:


1) Only having an `Xdawn` class accepting both numpy arrays and epoch, and no `XdawnTransformer`. This implementation becomes a nightmare to integrate with sklearn, 

* because `filters_` and `patterns_` are dict whose keys depend on `event_id`, which would thus need to pass as extra param in the init. 

* `correct_overlap` would only work for epochs. IMO this is bad design because the function would vary as function of input type.


2) `XdawnTransformer` accepting both epochs and numpy arrays: i.e. `xd.fit(epochs)` or `xd.fit(X, y)`: This implementation is complex as we would need to 

* pick the data channels and exclude the stim, eog etc

* check that its the same picking at `fit()` and `transform()`

* input epochs but still output numpy arrays, which is weird. We could add an `output='epochs'` parameter, but it doesn't seem the right way to go.


3) `XdawnTransformer` only accepting numpy arrays: `xd.fit(X, y)`, and keeping the current `Xdawn`. The main fitting function is currently private, but could be made public, so that advanced users can  apply an overlap correction.

Additionally, I:

* corrected a few bugs in the Xdawn class which was storing unnecessary or undocumented attributes (`xd.overlap_`, `xd.event_id_` and remove `xd.exclude`).

* removed the `mne/decoding/xdawn.py` and put both `Xdawn` and `XdawnTransformer` into `mne/preprocessing/xdawn.py`

However, I did not completely refactored the Xdawn class. The `apply()` code is already quite messy and I would be strongly in favour of deprecating the class entirely as:

* Xdawn is applied across different channels types, without rescaling, which doesn't seem right to me. Different channel have different scalings, and will thus differentially contribute to the eigen values.

* Xdawn fits the maximum number of components even if `n_components` is low, but removes the excluded components at `apply()`. This API does not follow the sklearn standard of setting the parameters at `init()`, and is not computationally efficient. I understand that this is for user to try to see how many components they should remove to clean their data, but this is really pushing for ad-hoc parameterizations and overfitting.

* The behavior of `Xdawn.apply(exclude=exclude)` is not very clear to me. The code suggests that the excluded components are not solely defined by `n_components` (defined at init) or the optional `exclude` parameter (defined at apply), but by a combination of the two.


So, ATM, I have not deprecated Xdawn as I suspect @agramfort wants a high level class, but I can already say that I don't think I won't be the one supporting such high level beast in the future.

Eric89GXL commented:
@kingjr do you see this as being mergeable today? If not, is there any way you could split out the fix or failing `master` builds (#3433) into a separate PR?

kingjr commented:
I'm only waiting for reviews

jona-sassenhagen commented on examples/preprocessing/plot_xdawn_denoising.py:
```diff
@@ -76,5 +75,5 @@
 epochs_denoised = xd.apply(epochs)
 
-# Plot image epoch after xdawn
+# Plot
```
Why not a more descriptive one?

Eric89GXL commented:
Can you set to MRG then?

To ask a slightly different, question do you expect this *to be merged* today? Waiting for reviews and API comments could mean several more iterations, and more days of waiting, especially since current holiday time might make the review lag long. Based on the length of your last comment, I'm assuming this will be the case, no?

In the meantime we should try to get `master` passing ASAP -- is it easy to split off a simple fix for it?

jona-sassenhagen commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -29,54 +30,55 @@ def _get_data():
-    # with this parameters, the overlapp correction must be False
-    assert_equal(xd.correct_overlap, False)
-    # no overlapp correction should give averaged evoked
+    # With this parameters, the overlap correction must be False
```
these parameter**s**

kingjr commented:
ok, I'll try to give a quick fix to the example in a separate PR.

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,362 @@
+    # Contruct raw signal
+    raw = _construct_signal_from_epochs(epochs_data, events, sfreq, tmin)
+
+    # Compute average evoked
```
I'm not sure that's the correct term considering the overlap correction. Maybe more something like "Compute the independent evoked responses per condition, correcting for overlap".

(No evokeds are ever averaged, and no averaging is actually done.)

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,362 @@
 
-    Returns
-    -------
-    evokeds : dict of evoked instance
```
So the evokeds are not returned anymore?

kingjr replied:
Well, I figured that it is not the role of the Xdawn algorithm per se. Its function is to find the spatial filters, not to store additional convenient variables. 

However, it is still output by the private function `_fit_xdawn()`, mainly to remain compatibly with the original `Xdawn`. Once the latter is deprecated, I would be in favour of not returning `evokeds` and make `fit_xdawn()` public.

WDYT?

jona-sassenhagen replied:
>However, it is still output by the private function _fit_xdawn()

Yes, I saw that later (see comment below). That's okay then.

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -271,17 +367,16 @@ class Xdawn(_Xdawn, ContainsMixin):
+    correct_overlap : 'auto' or bool (default 'auto')
         Apply correction for overlaped ERP for the estimation of evokeds
         responses. if 'auto', the overlapp correction is chosen in function
         of the events in epochs.events.
```
Mangled english

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -318,63 +417,82 @@ class Xdawn(_Xdawn, ContainsMixin):
+            n_events = len(epochs[eid])
+            evoked = EvokedArray(this_evo, epochs.info, tmin=epochs.tmin,
+                                 comment=eid, nave=n_events)
+            self.evokeds_[eid] = evoked
```
Ah I see. So this is where the evokeds are now?

kingjr replied:
This is the old `Xdawn` API. I'm for deprecating this class in favour of `XdawnTransformer`.

jona-sassenhagen replied:
So ... not having any way of directly retrieving the evokeds?

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,362 @@
+        evo = np.dot(evo, toeplitz)
+        evo_cov = np.matrix(_regularized_covariance(evo, reg))
+
+        # Spatial filters
```
Can you clarify if this *constructs* spatial filters or *applies* spatial filters? (I think the former right?)

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,362 @@
+            not np.array_equal(signal_cov.shape,
+                               np.tile(epochs_data.shape[1], 2))):
+        raise ValueError('signal_cov must be None, a covariance instance '
+                         'or a array of shape (n_chans, n_chans)')
```
**an** array

And I'd insert an Oxford comma after 'insance'.

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,362 @@
     # Concatenate toeplitz
-    full_toep = np.concatenate(full_toep)
+    toeplitz = np.array(toeplitz)
+    full_toep = np.concatenate(toeplitz)
```
Maybe just call it X and put in a comment that it's diagonals.

kingjr replied:
I actually don't understand this toep thing, so I kept the names. 

I'd rather not call it X to not confuse it with the sklearn X

jona-sassenhagen commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -86,67 +88,133 @@ def test_xdawn_fit():
+    # With shrinkage
     xd = Xdawn(n_components=2, correct_overlap=False,
                signal_cov=np.eye(len(picks)), reg=0.1)
     xd.fit(epochs)
```
Loop over reg types I guess. You only vary `reg=` after all?

jona-sassenhagen commented on mne/preprocessing/tests/test_xdawn.py:
```diff
@@ -86,67 +88,133 @@ def test_xdawn_fit():
     denoise = xd.apply(epochs)
-    # apply on evoked
+    # Apply on evoked
     xd.apply(epochs.average())
```
I guess here you could also loop over the possible input types.

jona-sassenhagen commented:
As you can see, I have a lot of comments, but they're about rather trivial stuff.

jona-sassenhagen commented:
I don't have time to look at the API today :(

jona-sassenhagen commented on mne/preprocessing/xdawn.py:
```diff
@@ -1,266 +1,362 @@
-    evoked_data = dict()
-    for idx, eid in enumerate(event_id):
-        evoked_data[eid] = all_evokeds[idx].T
+    evokeds_data = np.dot(predictor, raw.T)
```
Look here: it *is* X, this is simply constructing the pseudo inverse, and solving for the coefs in line 97.

It's literally just a least squares estimate.

```
pinv = np.dot(linalg.pinv(np.dot(X, X.T)), X)
np.dot(pinv, y)
```

jona-sassenhagen replied:
Not that important either way.

agramfort commented:
I am bit lost with this huge diff...

but travis is not happy ...

kingjr commented:
[Don't merge, I just realized that the overlap is not tested properly]

jona-sassenhagen commented:
Travis errors seem unrelated?

Eric89GXL commented:
One of them is (svd convergence) but there are some legit ones:
```
======================================================================
ERROR: Test Xdawn apply and transform.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/testenv/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/envs/testenv/lib/python3.5/site-packages/mne-0.13.dev0-py3.5.egg/mne/preprocessing/tests/test_xdawn.py", line 111, in test_xdawn_apply_transform
    xd.transform(epochs)
  File "/home/travis/miniconda/envs/testenv/lib/python3.5/site-packages/mne-0.13.dev0-py3.5.egg/mne/preprocessing/xdawn.py", line 519, in transform
    for filt in self.filters_.itervalues()]
AttributeError: 'dict' object has no attribute 'itervalues'
```

jona-sassenhagen commented:
JR get off Python 2 already :|

agramfort commented:
@kingjr you need to rebase

agramfort commented on examples/decoding/plot_decoding_xdawn_transformer.py:
```diff
@@ -17,7 +17,7 @@
 from mne import io, pick_types, read_events, Epochs
 from mne.datasets import sample
-from mne.decoding.xdawn import XdawnTransformer
+from mne.preprocessing.xdawn import XdawnTransformer
```
mne.preprocessing import XdawnTransformer

agramfort commented:
no need to update/fix what's new or reference.rst?

kingjr commented:
> no need to update/fix what's new or reference.rst?

It's only modifying underlying codes or the API of `XdawnTransformer` which wasn't in the previous release, so I don't think it's necessary.

kingjr commented:
@Eric89GXL there's still a linalg error, but I dont' know what to do with it. when run locally with 100x loop, it never crashes.

Eric89GXL commented:
It's a long-standing issue with Intel MKL. There is a workaround in SciPy 0.18 when it comes out. In the meantime we just have to live with it :(

kingjr commented:
@Eric89GXL ok, thanks for the precision.

@dengemann do you have time to give a review? Sorry, the PR is pretty dense.

agramfort commented:
travis is not happy

kingjr commented:
@agramfort it's a linalg problem, else it's happy

agramfort commented:
I cannot merge a PR with a broken travis related to the PR...

Eric89GXL commented:
It's not related as far as I can tell, it's the annoying mkl bug popping up
again. I guess I could open a PR into this branch to work around it?

agramfort commented:
if it makes travis green yes :)

Eric89GXL commented:
Actually, it looks like the error is no longer related to SVD in MaxFilter. @kingjr it is with some Xdawn test

kingjr commented:
@Eric89GXL it's a linalg error in the xdawn. I don't manage to replicate it locally. Let's see whether setting the  random seed fixes it.

Eric89GXL commented:
If not, I recently added the buggy_mkl test decorator

kingjr commented:
@Eric89GXL I'm stuck. I installed python 2.6 with scipy 11.0, numpy 1.7 etc, and I can't reproduce the linalg error. Any chance you could have a look?

Eric89GXL commented:
It is CPU architecture dependent use the decorator

agramfort commented:
@kingjr debug and use mkl debug decorator for the tests

kingjr commented:
> @kingjr debug 

Come on, that s not really helpful.

I'm already using the mkl decorator. See L88 of the test.

dengemann commented:
Sorry was offline, I will have time to review today and tomorrow.
On Mon, 25 Jul 2016 at 10:31, Jean-Rémi KING <notifications@github.com>
wrote:

> @kingjr <https://github.com/kingjr> debug
>
> Come on, that s not really helpful.
>
> I'm already using the mkl decorator. See L88 of the test.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3425#issuecomment-234886136>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB0finZDR1bqAFi9Fe0oVoRAaMyn1Pxaks5qZHRbgaJpZM4JN2vQ>
> .
>

Eric89GXL commented:
Your error is not the SVD error that I saw before, so the MKL decorator will not help. It is a different linear algebra error:

https://travis-ci.org/mne-tools/mne-python/jobs/147028330#L2833

It looks like it might be suggesting that your (covariance?) matrix is not positive definite. Are there zero eigenvalues? If so, it's not *positive definite* but *positive semidefinite*.

I googled the error and first result was something very similar:

https://github.com/scikit-learn/scikit-learn/issues/2640

I wonder if you should use some other linear algebra function like `svd` instead of `eigh`, since SVD won't care about the zero or negative eigenvalues.

Eric89GXL commented:
Without looking at the code, I'd assume that it's happening on element 60 because it's EEG data with 60 channels and an avg ref applied, i.e. of rank 59? That's one way this could happen.

kingjr commented:
I can relook into it, but I'm surprised it only crashes in one of the python job

Eric89GXL commented:
Slight numerical differences can crop up only sometimes. In this case they probably cause one of the eigenvalues to go to like -1e-16 or actual 0 instead of (in most cases) +1e-16. Or maybe `scipy` improved their check for numerical precision in the meantime, so you're hitting a version issue. In either case, it probably warrants further investigation unfortunately.

kingjr commented:
@Eric89GXL ok thanks. I've tried to use the same package versions as the failing job, without success. Let's see whether the explicit non reference makes it.

Eric89GXL commented:
Looks like it passed other than the unrelated flake8 package error

agramfort commented:
@kingjr ping us when CIs are happy

kingjr commented:
@agramfort IIUC Cis are happy, it's a non related flake8 bug that made it crash.

Eric89GXL commented:
@agramfort confirmed CIs are happy

agramfort commented:
thx @kingjr
